tests/cpp/test_memory.cpp

```diff
@@ -668,6 +679,89 @@ TEST_F(TMAIndexingTest, NonTrivialGmemAllocationDomain) {
 
 class TMAMiscTest : public TMATest {};
 
+TEST_F(TMAMiscTest, AdvancedThreadParallelizationLoad) {
+  GTEST_SKIP()
+      << "Advanced thread parallelization is not yet implemented. "
+         "Our current synchronization hacking is too naive to work here.";
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  const DataType dtype = DataType::Float;
+
+  auto tv0 = makeContigTensor(1, dtype);
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  fusion.addOutput(tv2);
+
+  tv1->setMemoryType(MemoryType::Shared);
+  tv1->definition()->as<LoadStoreOp>()->setOpType(
+      LoadStoreOpType::CpAsyncBulkTensorTile);
+
+  // Use 4 threads to issue TMA simultaneously
+  tv1->split(0, 128);
+  tv1->axis(1)->parallelize(ParallelType::Bulk);
+  tv1->split(0, 4);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+
+  // Use 512 threads to do the plain store simultaneously
+  tv2->split(0, 128);
+  tv2->split(0, 4);
+  tv2->merge(1);
+  tv2->axis(0)->parallelize(ParallelType::BIDx);
+  tv2->axis(1)->parallelize(ParallelType::TIDx);
+
+  auto options =
+      at::TensorOptions().dtype(data_type_to_aten(dtype)).device(at::kCUDA, 0);
+  auto t0 = at::randn({100000}, options);
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0}, {}, matmul_cparams);
+
+  auto cg_outputs = fe.runFusion({t0});
+  testValidate(&fusion, cg_outputs, {t0}, {t0}, __LINE__, __FILE__);
+}
+
+TEST_F(TMAMiscTest, AdvancedThreadParallelizationStore) {
```

Comment:
Kernel:

```CUDA
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  nvfuser_index_t i1;
  i1 = 512LL * ((nvfuser_index_t)blockIdx.x);
  nvfuser_index_t i2;
  i2 = ((nvfuser_index_t)threadIdx.x) + i1;
  float* T1 = reinterpret_cast<float*>(array + smem_offset + 0);
  if ((i2 < T0.logical_size[0LL])) {
    T1[((nvfuser_index_t)threadIdx.x)]
       = T0[i2];
  }
  __syncthreads();
  if ((((nvfuser_index_t)threadIdx.x) < 4)) {
    asm volatile("fence.proxy.async;\n");
    Hopper::cpAsyncBulkTensorTileS2G((Hopper::CpAsyncBulkTensorTileS2GIndex<1>{ (&var0), (Array<nvfuser_index_t, 1, 1>{((128LL * ((nvfuser_index_t)threadIdx.x)) + i1)}) }), (toSmem(T1) + (512 * ((nvfuser_index_t)threadIdx.x))));
    asm volatile("cp.async.bulk.commit_group;\n");
    asm volatile("cp.async.bulk.wait_group.read %0;\n"::"n"(0LL):"memory");
  }
}
```

Reply:
I will postpone reviewing store operation, I need to check what operands should be used for it.